Set up single-command install

Windyo opened the pull request:
With the help of @dhbahr :

- added sanity_checks.py which checks if setup commands have been run, and return booleans based on results
- added processing of sanity checks in webserver.sh, with exit conditions
- added example docker-compose file
- added stub in readme.md

I don't know if this needs more documentation or not. Closes #12

milesmcc commented:
Thanks for putting this together! I'll go through and make a few small changes this weekend before hopefully merging.

One important thing, though: can we move the migrations and sanity checks out of `webserver.sh`? I think `webserver.sh` needs to have as few side effects as possible. It needs to run the webserver—and nothing else. Maybe you could separate those checks into a separate `setup.py` file?

milesmcc commented on shynet/webserver.sh:
```diff
@@ -1,8 +1,51 @@
+  echo "Running Migrations..."
+  {
+    python manage.py migrate && echo "Migrations Done"
+  } || {
+    echo "Failed Migrations, exiting" && exit 1
+  }
+  else
+  echo "Migrations Unecessary, skipping"
+fi
+if [[ ${sanity_results[1]} == True ]]; then
+  echo "Running CreateAdmin..."
+  {
+    temppwd=$( python manage.py registeradmin $SHYNET_EMAIL ) && echo "Admin Created, password $temppwd"
+  } || {
+    echo "Failed CreateAdmin, exiting" & exit 1
+  }
+  else
+  echo "CreateAdmin Unecessary, skipping"
+fi
+if [[ ${sanity_results[2]} == True ]]; then
+  echo "Setting Hostname..."
+  {
+    python manage.py hostname $SHYNET_HOST && echo "Host Set"
+  } || {
+    echo "Failed setting Hostname, exiting" & exit 1
+  }
+  else
+  echo "Hostname Unecessary, skipping"
+fi
+if [[ ${sanity_results[3]} == True ]]; then
+  echo "Setting Hostname..."
+  {
+    python manage.py whitelabel $SHYNET_NAME && echo "WhiteLabel Set"
+  } || {
+    echo "Failed Migrations, exiting" & exit 1
+  }
+  else
+  echo "WhiteLabel Unecessary, skipping"
+fi
+
```
Can we move this into its own file? I'm concerned about putting this in `webserver.sh`. For multi-tenant systems, running migrations unexpectedly could have serious side-consequences.

Windyo replied:
We can, no issue, but technically it only does the migrations if the tables don't exist, not if migrations are needed. If migrations are needed but the tables exist, the admin will have to run the commands manually, which I think is normal as per your comment.

As an aside the only way to have this run on first run is to have it be in the docker entrypoint. A resolution to this would therefore be to have another .sh "entrypoint.sh" which calls the checks and then calls webserver.sh.

Otherwise we'd need a less lightweight solution which would be to actually set a flag in postgres on first run that says "first run done", so that the checks don't ever get evaluated again after init, but that sounds like overkill to me... Thoughts ?

milesmcc replied:
I like the idea of having two entry points. Does docker-compose allow specifying container commands? If so, we can leave the unmodified webserver.sh as the default CMD, and add another script that the compose file can use which would run these checks and then the webserver script. Does that work?

Windyo replied:
compose technically supports overriding the entrypoint, but only in the context of `run`, which is a one-time command.
`up` doesn't support entrypoint override, and is the standard way of running services with Compose.

The main "problem" here is that the shynet container itself is purely throwaway - it doesn't need to store anything persistent, that's what the postgre db is here for. So we have no volume to store the "yeah I ran once, don't run checks again and just call webserver.sh plz kthxbye" flag.
Now of course we could just create a volume for that, but that seems a bit... cumbersome, or store that flag in postgre itself, though that'll require a database callout to store the information of where the setup is (in case it gets interrupted).

The simplest way IMO is simply to:

- take migrations and spin it into its own migcheck.sh
- take the other checks and do "firstrun.sh" with it
- have an entrypoint.sh that:
	- runs firstrun.sh, and if any check comes back FALSE (meaning first run was done at least once), just calls webserver.sh
    - runs migcheck.sh only if all three checks before returned TRUE (meaning there's literally no admin user no nothing, so we can safely setup the database because it's empty)
    - uns webserver.sh as if both checks above see all's fine.

Windyo replied:
... or we expose an environment variable FIRST_RUN and tie the checks to that being set to true manually in the Compose file. I'm actually OK with that on a design perspective - it'll work even if you want to spawn multiple instances of shynet and be easily adaptable to swarm mode.

milesmcc replied:
Another option is to change the default entry point to some kind of “launch” script that does these checks, and keep “webserver.css” exposed for those who want the server to run without side effects? I’m not opposed to changing the default script, as long as we keep “webserver.sh” available.

Windyo replied:
that's what I meant with entrypoint.sh, lemme whip that up for ya, see if it matches your expectations more :)

milesmcc replied:
Ah got it, sounds good!

milesmcc commented on shynet/core/management/commands/sanity_checks.py:
```diff
@@ -0,0 +1,49 @@
+
+
+class Command(BaseCommand):
+    help = "Performs safe checks on the Shynet setup"
```
Grammar fix; can we change to "Performs sanity checks on the Shynet setup"?

Windyo replied:
done

milesmcc commented on README.md:
```diff
@@ -93,7 +93,11 @@ Shynet is pretty simple, but there are a few key terms you need to know in order
 ## Installation
 
-You can find installation instructions in the [Getting Started Guide](GUIDE.md#installation).
+If you use docker-compose, you can use the docker-compose.yml file to set up shynet and launch it by doing docker-compose up.
+Note that the application will generate a password for you on initial run, which will be output to standard docker logs.
+Also note that this compose file assumes you use a reverse-proxy in front of it, whether Traefik v2, Nginx-Proxy, or a solution on your host.
+
+You can find further installation instructions in the [Getting Started Guide](GUIDE.md#installation).
```
I'm going to rephrase this a bit to better fit the way the rest of the documentation is written, but I'll do this on my end before merging.

milesmcc commented:
Thanks for putting this together!

Windyo commented:
Squashed all commits, added Entrypoint.sh, removed commands from Webserver.sh, changed Dockerfile to reference new entrypoint, changed minor grammar as requested.

Let me know if that fits your vision more.
Cheers

milesmcc commented:
Looks good; I will do a final review and some small edits this weekend, then merge.

milesmcc commented:
@Windyo I've merged these changes onto the `dev` branch, pending merge. Would you mind taking one final look through to make sure everything works? I made a few changes (e.g. moving all environment variables into `TEMPLATE.md`) that might have broken things, and want to make sure everything looks good to you before I merge,

milesmcc commented:
Did a bit of testing myself, and everything seems to work well! Thanks @Windyo and @dhbahr for all of your hard work here.

Windyo commented:
Looks good to me :) sorry, was otherwise occupied.

I'll try to see if there's any way to lighten the image next, the 1.3GB weight seems weird to me for such an app.

milesmcc commented:
Great, thanks!
